transaction/core/src/input_rules.rs

```diff
@@ -53,6 +69,103 @@ impl InputRules {
                 return Err(InputRuleError::MissingRequiredOutput);
             }
         }
+        // Partial-Fill rules verification (MCIP #42)
+        if let Some(fractional_change) = self.fractional_change.as_ref() {
+            // There is a fractional change output. Let's try to unblind its amount.
+            let fractional_change_amount = fractional_change.reveal_amount()?;
+            // Let's check if there is a corresponding real change output.
+            let real_change = tx
+                .prefix
+                .outputs
+                .iter()
+                .find(|x| {
+                    x.public_key == fractional_change.tx_out.public_key
+                        && x.target_key == fractional_change.tx_out.target_key
+                })
+                .ok_or(InputRuleError::MissingRealChangeOutput)?;
+            // Let's try to unblind its amount.
+            let real_change_amount =
+                try_reveal_amount(real_change, fractional_change.amount_shared_secret.as_ref())?;
+
+            // Check the bounds of the real change amount
+            if real_change_amount.token_id != fractional_change_amount.token_id {
+                return Err(InputRuleError::RealOutputTokenIdMismatch);
+            }
+            if real_change_amount.value > fractional_change_amount.value {
+                return Err(InputRuleError::RealOutputAmountExceedsFractional);
+            }
+            if self.max_allowed_change_value != 0
+                && real_change_amount.value > self.max_allowed_change_value
+            {
+                return Err(InputRuleError::RealChangeOutputAmountExceedsLimit);
+            }
+
+            // Compute fill-fraction num and denom. This is the fraction of the maximum
+            // possible trade that could have occurred which did occur. This
+            // calculation is used later when checking if real output amounts respected the
+            // fill fraction.
+            let fill_fraction_num =
+                (fractional_change_amount.value - real_change_amount.value) as u128;
+            let fill_fraction_denom = fractional_change_amount.value as u128;
+
+            // Verify fractional_outputs
+            for fractional_output in self.fractional_outputs.iter() {
+                // Try to unblind the fractional output amount.
+                let fractional_output_amount = fractional_output.reveal_amount()?;
+                // Let's check if there is a corresponding real output.
+                let real_output = tx
+                    .prefix
+                    .outputs
+                    .iter()
+                    .find(|x| {
+                        x.public_key == fractional_output.tx_out.public_key
+                            && x.target_key == fractional_output.tx_out.target_key
```

Comment:
Coding it this way is a bit janky, it would be better if we had a special function on `TxOut` which means "check if a and b are equal EXCEPT in the masked_amount portion". That way if we add future fields to TxOut's they will still be enforced to match at this part

Reply:
Will fix shortly